### PR TITLE
ipn/{ipnlocal,ipnstate,localapi}: add localapi endpoints for client self-update

### PR DESCRIPTION
**Pull request by tendstofortytwo**
Updates #10187.

Some prior discussion done in #9930, this PR is a result of splitting that one in two.

### CODE REVIEW

**Comment by willnorris** [comment]
You can fix the depaware CI failure by running `make updatedeps`, and commit whatever changes that makes.

File: ipn/ipnstate/ipnstate.go [hunk 1]
```diff
@@ -710,3 +711,25 @@ type DebugDERPRegionReport struct {
 	Warnings []string
 	Errors   []string
 }
+
+type ProgressStatus string
```
Comment (willnorris) [inline ipn/ipnstate/ipnstate.go:715]: What about calling this `UpdateStatus` ?  I see that we could conceivably use this type for other, non-update statuses, but I suspect we would want those to be their own type?

File: ipn/localapi/localapi.go [hunk 2]
```diff
@@ -2325,6 +2329,71 @@ func (h *Handler) serveDebugLog(w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusNoContent)
 }
 
+// serveUpdateCheck returns the ClientVersion from Status, which contains
+// information on whether an update is available, and if so, what version,
+// *if* we support auto-updates on this platform. if we don't, this endpoint
+// always returns a ClientVersion saying we're running the newest version.
+// effectively, it tells us whether serveUpdateInstall will be able to install
```
Comment (willnorris) [inline ipn/localapi/localapi.go:2336]: s/effectively/Effectively/

File: ipn/localapi/localapi.go [hunk 3]
```diff
@@ -2325,6 +2329,71 @@ func (h *Handler) serveDebugLog(w http.ResponseWriter, r *http.Request) {
...
+			json.NewEncoder(w).Encode(tailcfg.ClientVersion{RunningLatest: true})
+		} else {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+		return
+	}
+
+	cv := h.b.StatusWithoutPeers().ClientVersion
+	if cv == nil {
+		cv = &tailcfg.ClientVersion{RunningLatest: true}
```
Comment (willnorris) [inline ipn/localapi/localapi.go:2360]: this confused me a bit.  Under what circumstances would `ClientVersion` be nil?  Add a comment explaining what's happening here

File: ipn/localapi/localapi.go [hunk 4]
```diff
@@ -2325,6 +2329,71 @@ func (h *Handler) serveDebugLog(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	json.NewEncoder(w).Encode(cv)
+}
+
+// serveUpdateInstall sends a request to the LocalBackend to start a Tailscale
+// self-update. it does not indicate whether or not the update or the request
```
Comment (willnorris) [inline ipn/localapi/localapi.go:2367]: This sentence reads a little funny.  Maybe...

> A successful response does not indicate whether the update succeeded, only that the request was accepted.

File: ipn/localapi/localapi.go [hunk 5]
```diff
@@ -2325,6 +2329,71 @@ func (h *Handler) serveDebugLog(w http.ResponseWriter, r *http.Request) {
...
+// to start it succeeded, only that it was accepted. clients should use
```
Comment (willnorris) [inline ipn/localapi/localapi.go:2368]: Clients  (same sentence capitalization throughout)

File: ipn/ipnlocal/local.go [hunk 6]
```diff
@@ -5492,6 +5497,54 @@ func (b *LocalBackend) DebugBreakDERPConns() error {
...
+		},
+	})
+	if err != nil {
+		b.pushSelfUpdateProgress(ipnstate.NewUpdateProgress(ipnstate.UpdateFailed, err.Error()))
+	}
+	err = up.Update()
+	if err != nil {
+		b.pushSelfUpdateProgress(ipnstate.NewUpdateProgress(ipnstate.UpdateFailed, err.Error()))
+	} else {
+		b.pushSelfUpdateProgress(ipnstate.NewUpdateProgress(ipnstate.UpdateFinished, "tailscaled did not restart, please restart Tailscale manually!"))
```
Comment (willnorris) [inline ipn/ipnlocal/local.go:5544]: I think a semi-colon is technically more appropriate than a comma in this case.  Also, probably no exclamation at the end... don't want to be yelling at users :)

**Review by willnorris — APPROVED** [review]
Several comments, though all relatively minor.

File: ipn/ipnstate/ipnstate.go (hunk 1 above)
Comment (tendstofortytwo) [reply ipn/ipnstate/ipnstate.go:715]: calling it `SelfUpdateStatus` then - both "update" and "status" are so generic words that I was feeling the same issue you described with `UpdateStatus`

File: ipn/localapi/localapi.go (hunk 2 above)
Comment (tendstofortytwo) [reply ipn/localapi/localapi.go:2336]: god I need to get back into capitalizing things

**Pull request merged**